src/dbnode/persist/fs/commitlog/commit_log.go

```diff
@@ -213,32 +250,42 @@ func (l *commitLog) flushEvery(interval time.Duration) {
 		}
 
 		// Request a flush
-		l.RLock()
-		if l.closed {
-			l.RUnlock()
+		l.closedState.RLock()
+		if l.closedState.closed {
+			l.closedState.RUnlock()
 			return
 		}
 
 		l.writes <- commitLogWrite{valueType: flushValueType}
-		l.RUnlock()
+		l.closedState.RUnlock()
 	}
 }
 
 func (l *commitLog) write() {
+	// This loop is the only part of the commit log that is allowed to modify (open, close, set to nil)
+	// the writer. As a result, it does not need to synchronize itself when it is using the writer (it
+	// is guaranteed to still be present), but it does need to acquire an exclusive lock when the writer
+	// is opened, closed, or set to nil (any operation that could change the pointer value of
+	// writerState.writer or writerState.activeFile). In other words, this function can be thought of
+	// as having an implied read lock at all times that is occasionally upgraded to an exclusive lock
+	// for the purpose of mutating the writerState.
 	for write := range l.writes {
 		// For writes requiring acks add to pending acks
 		if write.completionFn != nil {
-			l.pendingFlushFns = append(l.pendingFlushFns, write.completionFn)
+			l.flushState.pendingFlushFns = append(l.flushState.pendingFlushFns, write.completionFn)
```

Comment:
isn't this racy? i.e. you're modifying a field on `flushState` without a write Lock on it. but you use a Read lock when accessing the same field on line 326.